### PR TITLE
Memory efficient xcontent filtering

### PR DESCRIPTION
**Pull request by nik9000**
I found myself needing support for something like `filter_path` on
`XContentParser`. It was simple enough to plug it in so I did. Then I
realized that it might offer more memory efficient source filtering
(#25168) so I put together a quick benchmark comparing the source
filtering that we do in `_search`. This doesn't actually make the change
to the source fetching phase - it just benchmarks how it'd be.

Filtering using the parser is about 33% faster than how we filter now
when you select a single field from a 300 byte document:
```
Benchmark                                          (excludes)  (includes)  (source)  Mode  Cnt     Score    Error  Units
FetchSourcePhaseBenchmark.filterObjects                           message     short  avgt    5  2360.342 ±  4.715  ns/op
FetchSourcePhaseBenchmark.filterXContentOnBuilder                 message     short  avgt    5  2010.278 ± 15.042  ns/op
FetchSourcePhaseBenchmark.filterXContentOnParser                  message     short  avgt    5  1588.446 ± 18.593  ns/op
```

The top line is the way we filter now. The middle line is adding a
filter to `XContentBuilder` - something we can do right now without any
of my plumbing work. The bottom line is filtering on the parser,
requiring all the new plumbing.

This isn't particularly impresive. 33% *sounds* great! But 700
nanoseconds per document isn't going to cut into anyone's search times.
If you fetch a thousand docuents that's .7 milliseconds of savings.

But we mostly advise folks to use source filtering on fetch when the
source is large and you only want a small part of it. So I tried when
the source is about 4.3kb and you want a single field:
```
Benchmark                                          (excludes)  (includes)      (source)  Mode  Cnt     Score     Error  Units
FetchSourcePhaseBenchmark.filterObjects                           message  one_4k_field  avgt    5  5957.128 ± 117.402  ns/op
FetchSourcePhaseBenchmark.filterXContentOnBuilder                 message  one_4k_field  avgt    5  4999.073 ±  96.003  ns/op
FetchSourcePhaseBenchmark.filterXContentonParser                  message  one_4k_field  avgt    5  3261.478 ±  48.879  ns/op
```

That's 45% faster. Put another way, 2.7 microseconds a document. Not
bad!

But have a look at how things come out when you want a single field from
a 4 *megabyte* document:
```
Benchmark                                          (excludes)  (includes)      (source)  Mode  Cnt        Score        Error  Units
FetchSourcePhaseBenchmark.filterObjects                           message  one_4m_field  avgt    5  8266343.036 ± 176197.077  ns/op
FetchSourcePhaseBenchmark.filterXContentOnBuilder                 message  one_4m_field  avgt    5  6227560.013 ±  68306.318  ns/op
FetchSourcePhaseBenchmark.filterXContentonParser                  message  one_4m_field  avgt    5  1617153.472 ±  80164.547  ns/op
```

These documents are very large. I've encountered documents like them in
real life, but they've always been the outlier for me. But a 6.5
millisecond per document savings ain't anything to sneeze at.

Take a look at what you get when I turn on gc metrics:
```
FetchSourcePhaseBenchmark.filterObjects                          message  one_4m_field  avgt    5   7036097.561 ±  84721.312   ns/op
FetchSourcePhaseBenchmark.filterObjects:·gc.alloc.rate           message  one_4m_field  avgt    5      2166.613 ±     25.975  MB/sec
FetchSourcePhaseBenchmark.filterXContentOnBuilder                message  one_4m_field  avgt    5   6104595.992 ±  55445.508   ns/op
FetchSourcePhaseBenchmark.filterXContentOnBuilder:·gc.alloc.rate message  one_4m_field  avgt    5      2496.978 ±     22.650  MB/sec
FetchSourcePhaseBenchmark.filterXContentonParser                 message  one_4m_field  avgt    5   1614980.846 ±  31716.956   ns/op
FetchSourcePhaseBenchmark.filterXContentonParser:·gc.alloc.rate  message  one_4m_field  avgt    5         1.755 ±      0.035  MB/sec
```

### CODE REVIEW

**Comment by nik9000** [comment]
I'm not sure where to go from here. I think the xcontent changes could be useful outside of source filtering and I don't think I really have time to land them into source filtering. As much as I love that sweet sweet memory savings I'd be worried about this for the same reasons that @tlrx were in https://github.com/elastic/elasticsearch/issues/25168#issuecomment-312817839. Though I believe the savings from using the filtering on the reader are more substantial than in the linked issue - he's solution over there lines up fairly closely with my `filterXContentOnBuilder` implementation.

Like he said there - things like highlighting cause you to lose all the savings. Well, most of the savings at least. We could probably make them ride along with the same filtering. But that feels quite complex.

**Comment by nik9000** [comment]
Keep in mind, these microbenchmarks look good *because* they are microbenchmarks. I think before we can really know what this'd do we'd have to run it in rally with realistic documents.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilterPathBasedFilter.java [hunk 1]
```diff
@@ -21,13 +21,21 @@
...
 
     /**
      * Marker value that should be used to indicate that none of the
      * property names/values matches one of the filter paths.
      */
     private static final TokenFilter NO_MATCHING = new TokenFilter() {
+        @Override
+        public String toString() {
+            return "NO_MATCHING";
+        }
```
Comment (nik9000) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/support/filtering/FilterPathBasedFilter.java:38]: I added these for easier debugging.

**Comment by nik9000** [comment]
Like the benchmark implies, I think you could use this during the source fetch phase to speed up the fetches.

I've noticed that you get invalid json when you have a `**` in your *exclude* filter. I imagine that if we plugged this into the fetch phase we'd want to either fix that or use the object based filtering in that case.

**Comment by nik9000** [comment]
I've marked this as `search/search` and `draft` because it is going in to do search things. It's fairly general though. I'm not sure.

File: server/src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractXContentFilteringTestCase.java [hunk 2]
```diff
@@ -47,8 +48,51 @@ private XContentBuilder createBuilder() throws IOException {
...
+        if (excludesFilter != null && Arrays.stream(excludesFilter).anyMatch(FilterPath::hasDoubleWildcard)) {
+            /*
+             * If there are any double wildcard filters the parser based
+             * filtering produced weird invalid json. Just field names
+             * and no objects?! Weird. Anyway, we can't use it.
+             */
+            return filterOnBuilder(sample, includes, excludes);
+        }
+        FilterPath[] includesFilter = FilterPath.compile(includes);
+        return filterOnParser(sample, includesFilter, excludesFilter);
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/common/xcontent/support/filtering/AbstractXContentFilteringTestCase.java:65]: The tests that this enables are super comprehensive. It's kind of a shame to do it randomly though. We could probably use more subclasses or parametersized tests or something to run them all non-randomly. They are plenty fast.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentParser.java [hunk 3]
```diff
@@ -39,6 +43,25 @@ public JsonXContentParser(NamedXContentRegistry xContentRegistry,
...
+        DeprecationHandler deprecationHandler,
+        JsonParser parser,
+        RestApiVersion restApiVersion,
+        FilterPath[] include,
+        FilterPath[] exclude
+    ) {
+        super(xContentRegistry, deprecationHandler, restApiVersion);
+        JsonParser filtered = parser;
+        if (exclude != null) {
+            filtered = new FilteringParserDelegate(filtered, new FilterPathBasedFilter(exclude, false), true, true);
```
Comment (nik9000) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentParser.java:57]: This should probably fail in `hasDoubleWildcard`. Unless I can figure out a way to make it work!

**Comment by nik9000** [comment]
The double wildcard on exclude case has been fixed in Jackson 7.13: https://github.com/FasterXML/jackson-core/issues/700

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentParser.java (hunk 3 above)
Comment (nik9000) [reply libs/x-content/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentParser.java:63]: I added this test.

**Comment by jaymode** [comment]
Drive by comment, this type of filtering may also exclude empty objects and arrays like what happens during response filtering, see #63842. I attempted to fix that but hit issues after opening the PR and never revisited it.

**Comment by nik9000** [comment]
> Drive by comment, this type of filtering may also exclude empty objects and arrays like what happens during response filtering, see #63842. I attempted to fix that but hit issues after opening the PR and never revisited it.

Thanks! I hadn't realized that was a thing. This doesn't actually plug it into the fetch phase - only demonstrate that it'd be faster. It also rigs up the plumbing that I want for tsdb, which is totally my ulterior motive for this. But I'll check if it filters out empty unrelated objects anyway.

**Comment by nik9000** [comment]
> > Drive by comment, this type of filtering may also exclude empty objects and arrays like what happens during response filtering, see #63842. I attempted to fix that but hit issues after opening the PR and never revisited it.
> 
> Thanks! I hadn't realized that was a thing. This doesn't actually plug it into the fetch phase - only demonstrate that it'd be faster. It also rigs up the plumbing that I want for tsdb, which is totally my ulterior motive for this. But I'll check if it filters out empty unrelated objects anyway.

I've double checked this - the map based filtering that we're using now always throws out empty arrays. This is what the things look like:
```
curl -uelastic:password -HContent-Type:application/json -XDELETE localhost:9200/test
curl -uelastic:password -HContent-Type:application/json -XPOST localhost:9200/test/_doc?refresh -d'{"array": []}'
curl -uelastic:password -HContent-Type:application/json -XPOST 'localhost:9200/test/_search?pretty&_source=-**.whatever'
...
    "hits" : [
      {
        "_index" : "test",
        "_id" : "3Fly0XsBELTu-qwH8ixM",
        "_score" : 1.0,
        "_source" : { }
      }
    ]
...
```

So I think this change can only help here.

**Review by romseygeek — APPROVED** [review]
LGTM

**Comment by nik9000** [comment]
Thanks @romseygeek! I've pulling in master and will merge when jenkins is happy.

For those following along at home this *doesn't* change the actual fetch phase - it *just* plumbs through the ability to filter an `XContentParser`, tests it, and adds a benchmark showing that it'd be wicked fast if we used it for the fetch phase. I need this off in tsdb land, but I

**Comment by nik9000** [comment]
The build failures want #77640 and #77643. Those are open but not yet merged.

**Pull request merged**